### ISSUE TITLE
Race condition between OnVisibleChanged and initialization leads to WebView2 not rendering

### ISSUE THREAD

**Issue by tiagoabreu** [MicrosoftEdge/WebView2Feedback#898]
**Description**
Sometimes, we found a race condition between the initialization of WebView2 and the `OnVisibleChanged` event handler.
The `OnVisibleChanged` event handler sets the property `IsVisible` of the private field `_coreWebView2Controller` only if it was already initialized. 
The problem is that if the initialization didn't finish when the `VisibleChanged` event triggered we will be left in a state where the WebView2 `Visible` property is set to `true` but the `IsVisible` property of `_coreWebView2Controller` will be false.

I believe this can be addressed by one of two ways:
- When initializing the `_coreWebView2Controller`, set the `IsVisible` property to the same of the WebView2
- An event handler on `CoreWebView2InitializationCompleted` that sets the same property, as the point above.

It is not always reproducible, as it is dependent on what was processed first.

**Version**
SDK: 1.0.705.50
Runtime: WebView2 Runtime 88.0.705.62
Framework: WinForms
OS: Win10, Office 365

We have an workaround for it, we added an event handler on `CoreWebView2InitializationCompleted` that basically does:
```
webview2.Visible = false;
webview2.Visible = true;
```
This forces the OnVisibleChanged event handler to run and correctly set the property of `_coreWebView2Controller`

[AB#31691940](https://microsoft.visualstudio.com/90b2a23c-cab8-4e7c-90e7-a977f32c1f5d/_workitems/edit/31691940)

**Comment by champnic** [comment]
Thanks for catching this. I've filed a bug on our backlog and we'll get this fixed!

**Comment by johna-ms** [comment]
Hi

After looking into this, I do not think the issue is a race. I'll explain below, but first I think the issue is an old bug on us with how we check whether the webview2 control is initialized. We make this initialization check before setting the controller's IsVisible property so if it was buggy, this could explain your behavior. The fix for this should be in the latest prerelease SDK 824. Please try updating and see if the issue persists.

The reason I'm doubtful of a race is that we initialize the controller and set webview2 control properties on the same thread. If the controller's initialization is valid, then any bug would be around ordering fragility ie. controller properties maybe inconsistent with control properties depending on when properties are set. However, the controller's IsVisible property defaults to the visibility of the parent window which in this case is the webview2 control's underlying window ([see this property for more info](https://docs.microsoft.com/en-us/dotnet/api/system.windows.forms.control.handle?view=net-5.0)). So IsVisible should be robust to ordering.

If any of this doesn't seem right or make sense, please let me know

**Comment by johna-ms** [comment]
@tiagoabreu could you please check and see if this issue repros on SDK prerelease 824?

**Comment by tiagoabreu** [comment]
@johna-ms I could replicate the issue on the SDK prerelease 824.

**Comment by johna-ms** [comment]
@tiagoabreu Could you give more information on how to repro this?

**Comment by johna-ms** [comment]
Hi @tiagoabreu I learned that a more recent change went in that should fix this. Could you please try 865-prerelease? And if that doesn't work please give more detailed info on how to repro so we can investigate

**Comment by tiagoabreu** [comment]
Hello @johna-ms, unfortunately I can still reproduce it on version 865-prerelease.

Unfortunately I cannot share the code where I can reproduce it. Using a VSTO add-in with backstage enabled, I open the office application and then quit the backstage. The taskpane opens as soon as I leave the backstage (it is tied to the PresentationOpen event in case of Powerpoint) and then I see the task pane plain white. Further openings of the task pane always work. 

The same behavior with the workaround I mentioned in the original post makes this bug not reproducible.

**Comment by tiagoabreu** [comment]
I looked more into this and was able to reproduce it in a sample solution. 
[WordWebView2AddIn.zip](https://github.com/MicrosoftEdge/WebView2Feedback/files/6414744/WordWebView2AddIn.zip)

To reproduce it you need to have the backstage enabled when opening the office application.

**Comment by johna-ms** [comment]
Thanks for this repro! I'm able to repro it on my side so I'll investigate.

**Comment by johna-ms** [comment]
Thanks for that repro! So this issue and [this issue 1077](https://github.com/MicrosoftEdge/WebView2Feedback/issues/1077) actually share the same root cause. The thread is somewhat long so I'll summarize here, but tl;dr this has been fixed for the upcoming release. I realize I've said this a couple of times, but this time I've verified with your repro app!

The issue is indeed around the CoreWebView2Controller's visibility. The controller's default visibility is that of the parent window - not the control, but the window that underpins it. Word loads the add-in containing the webview control while in backstage mode. The parent *window* is invisible here since what we are actually seeing is the backstage menu. But the *control*'s Visible property is set to true for reasons known to the Winforms framework. This is the reason for the incongruity, not a race. The fix is pretty simple, when we initialize the webview controller, we set the controller's IsVisible property to that of the control's. Fix should be in the next prerelease, glad you have a workaround, and appreciate your patience. Let me know if any of this is unclear!
